### DISCUSSION TITLE
comfortly "with" for Connection()

### DISCUSSION THREAD

**Discussion by dashedman** [PyMySQL/PyMySQL#946] [Ideas]
```python
class Connection():
    ...
    def __enter__(self):
        if not self.open:
            self.ping(reconnect=True)
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()
        if exc_val:
            raise
            
            
conn = Connection(...)

#for every request use
with conn, conn.cursor() as cur:
    ... #first request
    
...
time.sleep(many) #many times without requests
...

with conn, conn.cursor() as cur:
    ... #second request

# etc
```

if db drop the connection, context manager `with` do a reconnect in `__enter__()`

**Comment by methane** [comment]
What is the difference from this?

```py
def create_connection():
    return pymysql.connect(...)

with create_connection() as conn:
    # first use

with create_connection() as conn:
    # second use
```

I don't want to implement "maybe convenient for some people" behaviors.  Such behaviors make long-term maintenance cost. Convenient features must be implemented in the higher layer as possible. Low-level library must focus on low-level.

**Reply by dashedman** [reply]
So, why you adding implement for "with" without proffit? Just for ignore of exceptions? - it's bad behavior for low-level library  :c

**Reply by methane** [reply]
> So, why you adding implement for "with" without profit? 

Have you ever used file objects?  Using `with` for closing file object is best practice. And file objects don't provide "implicit reopen" as you suggested:

```py
f = open(path)

with f:
    d1 = f.read()

with f:  # open again; this is not supported.
    d2 = f.read()
```

Do you think it is without profit? It's bad behavior?

**Reply by dashedman** [reply]
It's not a low-level library)

I undestand you. 
In pep 343 -- The "with" Statement(Caching Context Managers), says "Many context managers (such as files and generator-based contexts) will be single-use objects." Many but not all...

I'll close this disscusion
